### ISSUE TITLE
Add `vim.api.nvim_buf_detach`

### ISSUE THREAD

**Issue by noib3** [neovim/neovim#17874]
### Feature already in Vim?

No

### Feature description

The current way to detach from a buffer attached via `nvim_buf_attach` seems to return `true` from one of the callbacks passed to it, as described in `:h api-lua-detach`.

This is a bit awkward when detaching in response to an autocmd event or a user command, for at least a couple of reasons:

  * the buffers are detached not when the event occurs, but the next time one of the callbacks is called. This introduces the need for some global state, something like a `buffers_to_be_detached` table that has to be persisted;
  
  * the value of that state has to be checked every time one of the attached callbacks is executed to know if it should detach or continue as usual, introducing some (probably minimal, but still annoying) unnecessary overhead;
  
Something like a `vim.api.nvim_buf_detach` function taking a list of buffer numbers would solve both these problems. The help docs already mention such a function under `:h nvim_buf_detach`, but it's not actually present in the `vim.api` table.

**Comment by clason** [comment]
Yes, it's marked as `FUNC_API_REMOTE_ONLY` in the code, so it's exposed only via RPC.

@bfredl will know what the rationale for this is.

**Comment by bfredl** [comment]
The rationale is that the current solution actually works, while the proposed alternative here does not.

**Comment by noib3** [comment]
@bfredl Could you elaborate on why that solution (or something similar that doesn't require to have state lying around) couldn't work? Is it because of how the code is structured internally?

If that function is already exposed to external plugins why can't it be added for in-process ones as well?

**Comment by bfredl** [comment]
Perhaps we could retrofit (optional) namespace id as identification of the callback. It would require a separate detachment function as the signature is different, tho.

**Comment by justinmk** [comment]
Instead of this, we should reimplement buffer-updates as regular events ("autocmds"). We ever introduced the `buf_attach` concept because of an arbitrary separate of "rpc events" vs "editor events", but there is no real reason for that, and we end up reimplementing all the scope/instrumention/etc features of one in the other.
